### PR TITLE
Add timeout argument to ExecInContainer

### PR DESCRIPTION
**Pull request by rhcarvalho**
**What this PR does / why we need it**: This is related to https://github.com/kubernetes/kubernetes/issues/26895. It brings a timeout to the signature of `ExecInContainer` so that we can take timeouts into account in the future. Unlike my first attempt in https://github.com/kubernetes/kubernetes/pull/27956, it doesn't immediately observe the timeout, because it is impossible to do it with the current state of the Docker Remote API (the default exec handler implementation).

**Special notes for your reviewer**: This shares commits with https://github.com/kubernetes/kubernetes/pull/27956, but without some of them that have more controversial implications (actually supporting the timeouts). The original PR shall be closed in the current state to preserve the history (instead of dropping commits in that PR).

Pinging the original people working on this change: @ncdc @sttts @vishh @dims 

**Release note**:

``` release-note
NONE
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33366)

### CODE REVIEW

File: pkg/kubelet/kubelet.go
Comment (rhcarvalho) [inline pkg/kubelet/kubelet.go]: Definitely, thanks for spotting it.

File: pkg/kubelet/kubelet.go
Comment (rhcarvalho) [inline pkg/kubelet/kubelet.go]: This PR is not changing `RunInContainer`, therefore this is calling `kl.runner.ExecInContainer` without setting a timeout. I didn't look at what uses `RunInContainer`, but perhaps it would benefit from timeouts (and the resize channel?).

File: pkg/kubelet/lifecycle/handlers.go
Comment (rhcarvalho) [inline pkg/kubelet/lifecycle/handlers.go]: Same story here, no notion of timeouts nor resize, etc.

File: pkg/kubelet/server/remotecommand/exec.go
Comment (rhcarvalho) [inline pkg/kubelet/server/remotecommand/exec.go]: `ServeExec` also misses the notion of timeout.

File: pkg/kubelet/lifecycle/handlers.go
Comment (ncdc) [inline pkg/kubelet/lifecycle/handlers.go]: This is for running lifecycle hooks. There is no user input.

File: pkg/kubelet/kubelet.go
Comment (rhcarvalho) [inline pkg/kubelet/kubelet.go]: Do you think we should go add another argument to `RunInContainer`?
If that method is "the old way", I would just keep it as is, no timeout support.

File: pkg/kubelet/server/remotecommand/exec.go
Comment (ncdc) [inline pkg/kubelet/server/remotecommand/exec.go]: If you wanted to support it, you would probably want to add it as a new request parameter and have the client send the desired timeout.

File: pkg/kubelet/kubelet.go
Comment (ncdc) [inline pkg/kubelet/kubelet.go]: There's no harm in it, if sig node agrees.

**Comment by rhcarvalho** [comment]
@ncdc when you were adding the `resize` argument, did you have the same feeling than I that the argument list is getting too long?

File: pkg/kubelet/server/remotecommand/exec.go
Comment (rhcarvalho) [inline pkg/kubelet/server/remotecommand/exec.go]: I'll leave that to an independent feature request and keep this focused on the exec probes.

**Comment by ncdc** [comment]
Yes

**Comment by rhcarvalho** [comment]
Rebased after #33432 moved code from/to:
- `pkg/kubelet/kubelet.go` -> `pkg/kubelet/kubelet_pods.go`
- `pkg/kubelet/kubelet_test.go` -> `pkg/kubelet/kubelet_pods_test.go`

@ncdc @sttts all comments are addressed, do we need someone else's approval here?

**Comment by yujuhong** [comment]
@rhcarvalho what's the rush of doing this if the default docker exec doesn't support this yet, and the e2e test you added is not run anywhere?

/cc @timstclair

**Comment by vishh** [comment]
This PR helps avoid stray execs from causing trouble. It helps with system
reliability overall.
I do agree that we should have docker support exec timeouts if we can.

**Comment by ncdc** [comment]
> I do agree that we should have docker support exec timeouts if we can.

That will require a code change in Docker

**Comment by yujuhong** [comment]
I wasn't questioning the value of having exec timeouts, but was trying to see how much value this PR brings right now, given docker doesn't support it.

File: pkg/kubelet/rkt/rkt.go
Comment (yujuhong) [inline pkg/kubelet/rkt/rkt.go]: /cc @yifan-gu

Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: I am OK with this if we have a plan on how to actually implement the timeout.

Comment (rhcarvalho) [inline pkg/kubelet/rkt/rkt.go]: For rkt it is possible to implement the timeout today -- refer to #27956, in particular https://github.com/kubernetes/kubernetes/pull/27956/commits/657cb46121c183e35bc0acc785e86fa201653724

Same for the nsenter backend: https://github.com/kubernetes/kubernetes/pull/27956/commits/d82bc859c8ffd612d0ec7c8eb4604d33f2502705

**Comment by rhcarvalho** [comment]
@yujuhong having a timeout argument in the ExecInContainer signature will hopefully help guiding whoever is working on the CRI implementation, to make sure it includes timeout support (unlike the case in the Docker API). I think it is a necessary step to fix #26895.

I'd be okay if the [probe docs](http://kubernetes.io/docs/user-guide/pod-states/#container-probes) would mention that Exec probes ignore the timeout (the undocumented truth); or that we support timeouts when using `nsenter` and `rkt` and document that `nativeclient` (the default) doesn't support timeouts.

If there is agreement to support timeouts only in `nsenter` and `rkt` for now, I can bring in the commits that do that (https://github.com/kubernetes/kubernetes/pull/27956/commits/657cb46121c183e35bc0acc785e86fa201653724 and https://github.com/kubernetes/kubernetes/pull/27956/commits/d82bc859c8ffd612d0ec7c8eb4604d33f2502705).

Thanks!

**Comment by timstclair** [comment]
What's this PR waiting on? I added a timeout parameter to the CRI [`ExecSync`](https://github.com/kubernetes/kubernetes/blob/6f80ec91d663946e157fc20aa4a32ec166c91b15/pkg/kubelet/api/v1alpha1/runtime/api.proto#L638), and the changes in this PR would simplify the handling a bit.

**Comment by sttts** [comment]
@timstclair afaik nothing is blocking this. If it helps your CRI code, let's go ahead and get this merged.

**Comment by sttts** [comment]
@k8s-bot test this issue

**Comment by sttts** [comment]
@rhcarvalho there are compile errors

**Comment by rhcarvalho** [comment]
@sttts thanks, looks like yet another rebase needed :cry:

**Comment by rhcarvalho** [comment]
CLA signed

**Comment by rhcarvalho** [comment]
Found the issue: `pkg/kubelet/kubelet_pods.go` lost a `import "time"` that used to be there...

**Comment by rhcarvalho** [comment]
@k8s-bot test this issue

**Comment by sttts** [comment]
@k8s-bot test this issue

**Comment by rhcarvalho** [comment]
Still broken because...

```
*criAdapter does not implement "k8s.io/kubernetes/pkg/kubelet/server/remotecommand".Executor (wrong type for ExecInContainer method)
```

More types were added implementing that interface.

**Comment by rhcarvalho** [comment]
```diff
diff --git a/pkg/kubelet/server/streaming/server.go b/pkg/kubelet/server/streaming/server.go
index 994e6be..8a3a6b2 100644
--- a/pkg/kubelet/server/streaming/server.go
+++ b/pkg/kubelet/server/streaming/server.go
@@ -299,7 +299,7 @@ var _ remotecommand.Executor = &criAdapter{}
 var _ remotecommand.Attacher = &criAdapter{}
 var _ portforward.PortForwarder = &criAdapter{}
 
-func (a *criAdapter) ExecInContainer(podName string, podUID types.UID, container string, cmd []string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan term.Size) error {
+func (a *criAdapter) ExecInContainer(podName string, podUID types.UID, container string, cmd []string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan term.Size, timeout time.Duration) error {
        return a.Exec(container, cmd, in, out, err, tty, resize)
 }
 ```

**Comment by sttts** [comment]
@k8s-bot test this issue

**Comment by rhcarvalho** [comment]
And one more missing import in tests:

```diff
diff --git a/pkg/kubelet/kubelet_pods_test.go b/pkg/kubelet/kubelet_pods_test.go
index bca8c3c..d51287c 100644
--- a/pkg/kubelet/kubelet_pods_test.go
+++ b/pkg/kubelet/kubelet_pods_test.go
@@ -24,6 +24,7 @@ import (
        "net"
        "sort"
        "testing"
+       "time"
 
        "github.com/stretchr/testify/assert"
        "github.com/stretchr/testify/require"
```

**Comment by rhcarvalho** [comment]
Note: the missing imports and the one `ExecInContainer` missing the timeout argument today are not because I'm messy and the original commits were wrong, but because the files changed in master. The imports were removed, and a new `ExecInContainer` method was created.

**Comment by sttts** [comment]
@k8s-bot test this issue

**Comment by sttts** [comment]
@k8s-bot ok to test

**Comment by rhcarvalho** [comment]
@sttts @ncdc @timstclair all green.

File: pkg/kubelet/lifecycle/handlers.go
Comment (timstclair) [inline pkg/kubelet/lifecycle/handlers.go]: I feel like we should probably have a timeout on this, but it's outside the scope of this PR. Maybe add a TODO (with an associated issue).

File: pkg/kubelet/kubelet_pods.go
Comment (timstclair) [inline pkg/kubelet/kubelet_pods.go]: *Aside:* Do you know what `RunInContainer` is used for? I feel like this should probably have a timeout too, or maybe the method should just be deprecated.

**Review by timstclair — APPROVED** [review]
LGTM, just a couple optional comments.

**Comment by rhcarvalho** [comment]
Rebased after #35661.

File: pkg/kubelet/kubelet_pods.go
Comment (rhcarvalho) [inline pkg/kubelet/kubelet_pods.go]: > Do you know what RunInContainer is used for?

Not really, at least not "historically speaking". It seems it is only used in https://github.com/kubernetes/kubernetes/blob/e4b41dd2920eb98402d745931964f11279a9ff2e/pkg/kubelet/server/server.go#L624 apart from tests. That can probably be refactored to eliminate the `RunInContainer` method.

File: pkg/kubelet/lifecycle/handlers.go
Comment (rhcarvalho) [inline pkg/kubelet/lifecycle/handlers.go]: @timstclair quoting a comment from @ncdc on Sep 26:

> This is for running lifecycle hooks. There is no user input.

In any case, I added the TODO as you asked, so you can investigate that later.

**Comment by rhcarvalho** [comment]
I don't understand what the [GKE smoke e2e failure](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/33366/pull-kubernetes-e2e-gke/2421/) means:

image

**Comment by yujuhong** [comment]
@k8s-bot cvm gke e2e test this

**Comment by timstclair** [comment]
LGTM, please squash the commits.

**Comment by rhcarvalho** [comment]
@timstclair squashed.

**Comment by rhcarvalho** [comment]
[Integration tests failure](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/33366/pull-kubernetes-unit/2575/) looks like a flake, many issues related to missing files: "executable file not found in $PATH", "open /etc/kubernetes/ca/ca.pem: no such file or directory", ...

**Comment by sttts** [comment]
@k8s-bot unit test this

**Comment by rhcarvalho** [comment]
The integration tests are flaking as in https://github.com/kubernetes/kubernetes/issues/25363

**Comment by ncdc** [comment]
@k8s-bot unit test this

**Comment by rhcarvalho** [comment]
@timstclair @ncdc @sttts do we need something else here to trigger a merge?

**Comment by ncdc** [comment]
It's in the queue - just wait 😄 http://submit-queue.k8s.io/#/queue

**Comment by rhcarvalho** [comment]
There we go again, one more rebase after https://github.com/kubernetes/kubernetes/pull/36020.

@timstclair PTAL

**Comment by rhcarvalho** [comment]
@timstclair I suspect your PR went before this in the merge queue because of the milestone? This waited for days in the queue and was not merged.

**Comment by rhcarvalho** [comment]
Missed a `timeout time.Duration` argument in `pkg/kubelet/lifecycle/handlers_test.go`, fixed.

**Comment by rhcarvalho** [comment]
All tests green again.

**Comment by timstclair** [comment]
> @timstclair I suspect your PR went before this in the merge queue because of the milestone? This waited for days in the queue and was not merged.

Correct. Thanks for handling the rebase.

**Comment by saad-ali** [comment]
Release-czar approved post-code freeze merge--This was LGTMed and in the merge-queue at code freeze time for 1.5. Adding 1.5 milestone to let it gets merged after code freeze.

**Pull request merged**
